### PR TITLE
[script] replace python script with scala

### PR DESCRIPTION
**Pull request by Avimitin**
This PR add a new Scala executable t1-helper to replace the run-test.py script.

### CODE REVIEW

**Review by sequencer — CHANGES_REQUESTED** [review]

File: build.sc [hunk 1]
```diff
@@ -17,6 +17,7 @@ import $file.common
 
 object v {
   val scala = "2.13.12"
+  val scala3 = "3.3.3"
```
Comment (sequencer) [inline build.sc:20]: this is unused

**Review by sequencer — CHANGES_REQUESTED** [review]
For scala3, I prefer removing brackets for making it more python-like

**Comment by Avimitin** [comment]
Please review the latest coding style.

File: script/src/Main.scala [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+package org.chipsalliance.t1.script
+
+import mainargs.{main, arg, ParserForMethods, Leftover, Flag, TokensReader}
+import scala.io.AnsiColor._
+
+object Logger {
+  def info(message: String) = {
+    println(s"${BOLD}${GREEN}[INFO]${RESET} ${message}")
+  }
+}
```
Comment (sequencer) [inline script/src/Main.scala:13]:
```suggestion
object Logger:
  def info(message: String) = println(s"${BOLD}${GREEN}[INFO]${RESET} ${message}")
```

File: script/src/Main.scala [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+  implicit object PathRead extends TokensReader.Simple[os.Path]:
+    def shortName = "path"
+
+    def read(strs: Seq[String]) = Right(os.Path(strs.head, os.pwd))
+
+  def resolveTestElfPath(
+      config: String,
+      caseName: String,
+      forceX86: Boolean = false
+  ): os.Path = {
```
Comment (sequencer) [inline script/src/Main.scala:25]:
```suggestion
  ): os.Path =
```

File: script/src/Main.scala [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        s".#t1.${config}.${caseAttrRoot}.${caseName}"
+      )
+      Logger.info(
+        s"Running `${nixArgs.mkString(" ")}` to get test case ELF file"
+      )
+      os.Path(os.proc(nixArgs).call().out.trim()) / "bin" / s"${caseName}.elf"
+
+    Logger.info(s"Using test ELF: ${finalPath}")
+    finalPath
+  }
```
Comment (sequencer) [inline script/src/Main.scala:47]:
```suggestion
```

File: script/src/Main.scala [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+
+  def resolveEmulatorPath(
+      config: String,
+      emuType: String,
+      isTrace: Boolean = false
+  ): os.Path = {
```
Comment (sequencer) [inline script/src/Main.scala:53]:
```suggestion
  ): os.Path =
```

File: script/src/Main.scala [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+      "--no-warn-dirty",
+      s".#t1.${config}.${target}"
+    )
+    Logger.info(s"Running `${nixArgs.mkString(" ")}` to get emulator")
+
+    val finalPath = os.Path(os.proc(nixArgs).call().out.trim()) / "bin" / "emulator"
+    Logger.info(s"Using emulator: ${finalPath}")
+
+    finalPath
+  }
```
Comment (sequencer) [inline script/src/Main.scala:69]:
```suggestion
```

File: script/src/Main.scala [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+  def resolveElaborateConfig(
+      outputDir: os.Path,
+      configName: String
+  ): os.Path = {
```
Comment (sequencer) [inline script/src/Main.scala:74]:
```suggestion
  ): os.Path =
```

File: script/src/Main.scala [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+      "--",
+      configName,
+      "-t",
+      outputDir.toString
+    )
+    Logger.info(s"Runnning `${nixArgs.mkString(" ")}` to get config")
+    os.proc(nixArgs).call()
+
+    outputDir / "config.json"
+  }
```
Comment (sequencer) [inline script/src/Main.scala:95]:
```suggestion
```

File: script/src/Main.scala [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+  def prepareOutputDir(
+      outputDir: Option[String],
+      outputBaseDir: Option[String],
+      config: String,
+      emuType: String,
+      caseName: String
+  ): os.Path = {
```
Comment (sequencer) [inline script/src/Main.scala:103]:
```suggestion
  ): os.Path =
```

File: script/src/Main.scala [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+      if (outputBaseDir.isEmpty) then
+        os.pwd / "testrun" / s"${emuType}emu" / config / pathTail
+      else
+        os.Path(outputBaseDir.get, os.pwd) / config / pathTail
+    else
+      os.Path(outputDir.get)
+
+    os.makeDir.all(path)
+    path
+  }
```
Comment (sequencer) [inline script/src/Main.scala:124]:
```suggestion
```

File: script/src/Main.scala [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+      ) forceX86: Boolean = false,
+      @arg(
+        name = "dump-cycle",
+        doc = "Specify the dump starting point"
+      ) dumpCycle: Int = 0,
+      @arg(
+        name = "cosim-timeout",
+        doc = "specify timeout cycle for cosim"
+      ) cosimTimeout: Int = 400000
+  ) = {
```
Comment (sequencer) [inline script/src/Main.scala:196]:
```suggestion
  ) =
```

File: script/src/Main.scala [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+
+    Logger.info(s"Starting IP emulator: `${processArgs.mkString(" ")}`")
+    os.proc(processArgs).call()
+
+    if (!noFileLog.value) then
+      Logger.info(s"Emulator log save to ${outputPath / "emulator.log"}")
+
+    if (trace.value) then
+      Logger.info(s"Trace file save to ${outputPath} / trace.fst")
+  }
```
Comment (sequencer) [inline script/src/Main.scala:265]:
```suggestion
```

File: script/src/Main.scala [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+      ) outDir: Option[String],
+      @arg(
+        name = "base-out-dir",
+        doc = "save result files in {base_out_dir}/{config}/{case}/{run_config}"
+      ) baseOutDir: Option[String],
+      @arg(
+        name = "emulator-path",
+        doc = "path to emulator"
+      ) emulatorPath: Option[String]
+  ) = {
```
Comment (sequencer) [inline script/src/Main.scala:302]:
```suggestion
  ) =
```

File: script/src/Main.scala [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            if (traceFile.isDefined) then
+              os.Path(traceFile.get, os.pwd)
+            else outputPath / "trace.fst"
+          }")
+      )
+    os.proc(emuArgs).call()
+
+    if (trace.value) then
+      Logger.info(s"Trace file save to ${outputPath} / trace.fst")
+  }
```
Comment (sequencer) [inline script/src/Main.scala:327]:
```suggestion
```

File: script/src/Main.scala [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+
+  @main def listConfig() = {
```
Comment (sequencer) [inline script/src/Main.scala:329]:
```suggestion
  @main def listConfig() =
```

File: script/src/Main.scala [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+      Seq(
+        "nix",
+        "run",
+        "--no-warn-dirty",
+        ".#t1.configgen",
+        "--",
+        "listConfigs"
+      )
+    ).call(cwd = os.pwd, stdout = os.Inherit, stderr = os.Inherit)
+  }
```
Comment (sequencer) [inline script/src/Main.scala:340]:
```suggestion
```

**Review by sequencer — COMMENTED** [review]
Do it in a Scala3(python) way :)

**Comment by Avimitin** [comment]
No I think it is more like [Ruby](https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html#the-end-marker-1), but better.

**Review by sequencer — APPROVED** [review]
In the next PR, we should remove original python script ;p

**Pull request merged**